README.md
# Elegant Flash Message Wrapper For Laravel


[![License](https://img.shields.io/github/license/nicoaudy/laravel-noty?style=flat-square)](https://packagist.org/packages/nicoaudy/noty)
[![Latest Version on Packagist](https://img.shields.io/packagist/v/nicoaudy/noty.svg?style=flat-square)](https://packagist.org/packages/nicoaudy/noty)
[![Total Downloads](https://img.shields.io/packagist/dt/nicoaudy/noty.svg?style=flat-square)](https://packagist.org/packages/nicoaudy/noty)

This is a elegant and configurable notify package to send flash messages in Laravel apps. A flash message is a message that is carried over to the next request by storing it in the session.


![Screenshot](./screenshots/noty-screenshot.png "Screenshot")


This is how it can be used:

## Installation

You can install the package via composer:

```bash
composer require nicoaudy/noty
```

You can publish the config file with:

```bash
php artisan vendor:publish --provider="Nicoaudy\Noty\NotyServiceProvider"
```

This is the contents of the published config file:

```php
return [
    /*
    * Title configuration color
    */
    'titleColor' => '',

    /*
    * Title configuration size
    */
    'titleSize' => '',

    /*
    * Title configuration line hight
    */
    'titleLineHeight' => '',

    /*
    * Message configuration color
    */
    'messageColor' => '',

    /*
    * Message configuration size
    */
    'messageSize' => '',

    /*
    * Message configuration line hight
    */
    'messageLineHeight' => '',

    /*
     * Positions:  bottomRight, bottomLeft, topRight, topLeft, topCenter, bottomCenter, center
     */
    'position' => 'topRight',

    /*
     * Themes:  light, dark
     */
    'theme' => 'light',

    /*
     * Layout:  1 (title inline with message), 2 (title on top of message)
     */
    'layout' => 1,

    /*
    * [integer|boolean] - Delay for closing event in milliseconds. Set false for sticky notifications.
    */
    'timeout' => 4000,

    /*
     * [boolean] - Displays a progress bar.
     */
    'progressBar' => true,

    /*
     * [string] - Progress bar color
     */
    'progressBarColor' => 'rgb(0, 255, 184)',

    /*
     * [string] - Progress bar linear / ease
     */
    'progressBarEasing' => 'linear',

    /*
     * [boolean] - Balloon
     */
    'balloon' => true,

    /*
     * [boolean] - enable to close?
     */
    'close' => true,

    /*
     * [boolean] - Close on click
     */
    'closeOnClick' => false,

    /*
     * [boolean] - Close on escape?
     */
    'closeOnEscape' => false,

    /*
     * Animate inside
     */
    'animateInside' => true, // bool

    /*
     * Enable drag for close
     */
    'drag' => true, // bool

    /*
     * Pause on hover
     */
    'pauseOnHover' => true, // bool

    /*
     * Reset on hover
     */
    'resetOnHover' => false, // bool

    /*
     * [string] - Transition In
     * Default toast open animation. It can be: bounceInLeft, bounceInRight, bounceInUp, bounceInDown, fadeIn, fadeInDown, fadeInUp, fadeInLeft, fadeInRight or flipInX.
     */
    'transitionIn' => 'fadeInUp',

    /*
     * [string] - Transition Out
     * Default toast close animation. It can be: fadeOut, fadeOutUp, fadeOutDown, fadeOutLeft, fadeOutRight, flipOutX
     */
    'transitionOut' => 'fadeOut',

    /*
     * [string] - Transition In Mobile
     */
    'transitionInMobile' => 'fadeInUp',

    /*
     * [string] - Transition Out Mobile
     */
    'transitionOutMobile' => 'fadeOutDown',
];
```

## Usage

Within your controllers, before you perform a redirect, make a call to the noty() function.

```php
public function store()
{
    noty()->flash('Cool', 'Your data has been saved');

    return home();
}
```

You may also do:

```php
noty()->flash('Hey!', 'This is message');
noty()->info('Hey!', 'This is message');
noty()->success('Hey!', 'This is message');
noty()->warning('Hey!', 'This is message');
noty()->error('Hey!', 'This is message');
```

With this message flashed to the session, you may now display it in your view(s). Because flash messages are so common, we provide a template out of the box to get you started. Place this before end of the body tag.

```php
    {{ noty_assets() }}
</body>
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email freek@spatie.be instead of using the issue tracker.

## Credits

-   [Nico Audy](https://github.com/NicoAudy)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `noty` will be documented in this file

## 1.0.0 - 2020-07-16

- initial release stable version

## 1.0.1 - 2020-07-16

- Update Readme


tests/TestCase.php
<?php

namespace Nicoaudy\Noty\Tests;

use Nicoaudy\Noty\NotyServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    public function setUp(): void
    {
        parent::setUp();

        $this->withFactories(__DIR__.'/database/factories');
    }

    protected function getPackageProviders($app)
    {
        return [
            NotyServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        $app['config']->set('database.default', 'sqlite');
        $app['config']->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);

        /*
        include_once __DIR__.'/../database/migrations/create_noty_table.php.stub';
        (new \CreatePackageTable())->up();
        */
    }
}


tests/ExampleTest.php
<?php

namespace Nicoaudy\Noty\Tests;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


src/Noty.php
<?php

namespace Nicoaudy\Noty;

use Illuminate\Session\Store;
use Illuminate\Support\Arr;

class Noty
{
    private $session;
    private $config;

    public function __construct(Store $session)
    {
        $this->session = $session;
        $this->config = config('noty');
    }

    public function config($config = [])
    {
        $this->session->flash('noty.config', $config);

        return $this;
    }

    public function info($title = null, $message = null)
    {
        return $this->message($message, 'blue', $title);
    }

    public function warning($title = null, $message = null)
    {
        return $this->message($message, 'yellow', $title);
    }

    public function success($title = null, $message = null)
    {
        return $this->message($message, 'green', $title);
    }

    public function danger($title = null, $message = null)
    {
        return $this->message($message, 'red', $title);
    }

    public function flash($title = null, $message = null)
    {
        return $this->message($message, 'primary', $title);
    }

    private function message($message, $type, $title)
    {
        $messages = [];

        if (! empty($message)) {
            if (! empty(session('noty.messages'))) {
                $messages = session('noty.messages');
            }

            if (! collect(Arr::flatten($messages))->contains($message)) {
                $messages[] = [
                    'text' => $message,
                    'type' => $type,
                    'title' => $title ?? null,
                ];
            }
        }

        $this->session->flash('noty.messages', $messages);
        $this->session->flash('noty.config', $this->config);

        $availTypes = ['blue', 'red', 'green', 'yellow', 'primary'];
        if (in_array($type, $availTypes)) {
            $this->session->flash('noty.config.type', $type);
        }

        return $this;
    }
}


src/NotyServiceProvider.php
<?php

namespace Nicoaudy\Noty;

use Illuminate\Support\ServiceProvider;

class NotyServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__ . '/../config/noty.php' => config_path('noty.php'),
            ], 'config');

            $this->publishes([
                __DIR__ . '/../public/css/toast.min.css' => public_path('vendor/noty/toast.min.css'),
                __DIR__ . '/../public/js/toast.min.js' => public_path('vendor/noty/toast.min.js'),
            ], 'assets');
        }

        $this->loadViewsFrom(__DIR__ . '/../resources/views', 'noty');
    }

    public function register()
    {
        $this->mergeConfigFrom(__DIR__ . '/../config/noty.php', 'noty');

        $this->app->singleton('noty', function () {
            return $this->app->make(Noty::class);
        });
    }
}


resources/views/noty.blade.php
<link rel="stylesheet" href="{{ asset('vendor/noty/toast.min.css') }}">
<script src="{{ asset('vendor/noty/toast.min.js') }}"></script>

@if(session()->has('noty.messages'))
    <script type="text/javascript">
        (function () {
			iziToast.settings({
				theme: '{{ (!empty(session('noty.config.theme')) ? session('noty.config.theme') : config('laravel-noty.theme'))  }}',
				titleColor: '{{ (!empty(session('noty.config.titleColor')) ? session('noty.config.titleColor') : config('laravel-noty.titleColor'))  }}',
				titleSize: '{{ (!empty(session('noty.config.titleSize')) ? session('noty.config.titleSize') : config('laravel-noty.titleSize'))  }}',
				titleLineHeight: '{{ (!empty(session('noty.config.titleLineHeight')) ? session('noty.config.titleLineHeight') : config('laravel-noty.titleLineHeight'))  }}',
				messageColor: '{{ (!empty(session('noty.config.messageColor')) ? session('noty.config.messageColor') : config('laravel-noty.messageColor'))  }}',
				messageSize: '{{ (!empty(session('noty.config.messageSize')) ? session('noty.config.messageSize') : config('laravel-noty.messageSize'))  }}',
				messageLineHeight: '{{ (!empty(session('noty.config.messageLineHeight')) ? session('noty.config.messageLineHeight') : config('laravel-noty.messageLineHeight'))  }}',
				position: '{{ (!empty(session('noty.config.position')) ? session('noty.config.position') : config('laravel-noty.position'))  }}',
				timeout: '{{ (!empty(session('noty.config.timeout')) ? session('noty.config.timeout') : config('laravel-noty.timeout'))  }}',
				layout: '{{ (!empty(session('noty.config.layout')) ? session('noty.config.layout') : config('laravel-noty.layout'))  }}',
				close: '{{ (!empty(session('noty.config.close')) ? session('noty.config.close') : config('laravel-noty.close'))  }}',
				closeOnClick: '{{ (!empty(session('noty.config.closeOnClick')) ? session('noty.config.closeOnClick') : config('laravel-noty.closeOnClick'))  }}',
				closeOnEscape: '{{ (!empty(session('noty.config.closeOnEscape')) ? session('noty.config.closeOnEscape') : config('laravel-noty.closeOnEscape'))  }}',
				animateInside: '{{ (!empty(session('noty.config.animateInside')) ? session('noty.config.animateInside') : config('laravel-noty.animateInside'))  }}',
				drag: '{{ (!empty(session('noty.config.drag')) ? session('noty.config.drag') : config('laravel-noty.drag'))  }}',
				pauseOnHover: '{{ (!empty(session('noty.config.pauseOnHover')) ? session('noty.config.pauseOnHover') : config('laravel-noty.pauseOnHover'))  }}',
				resetOnHover: '{{ (!empty(session('noty.config.resetOnHover')) ? session('noty.config.resetOnHover') : config('laravel-noty.resetOnHover'))  }}',
				progressBar: '{{ (!empty(session('noty.config.progressBar')) ? session('noty.config.progressBar') : config('laravel-noty.progressBar'))  }}',
				progressBarColor: '{{ (!empty(session('noty.config.progressBarColor')) ? session('noty.config.progressBarColor') : config('laravel-noty.progressBarColor'))  }}',
				progressBarEasing: '{{ (!empty(session('noty.config.progressBarEasing')) ? session('noty.config.progressBarEasing') : config('laravel-noty.progressBarEasing'))  }}',
				balloon: '{{ (!empty(session('noty.config.balloon')) ? session('noty.config.balloon') : config('laravel-noty.balloon'))  }}',
				transitionIn: '{{ (!empty(session('noty.config.transitionIn')) ? session('noty.config.transitionIn') : config('laravel-noty.transitionIn'))  }}',
				transitionOut: '{{ (!empty(session('noty.config.transitionOut')) ? session('noty.config.transitionOut') : config('laravel-noty.transitionOut'))  }}',
				transitionInMobile: '{{ (!empty(session('noty.config.transitionInMobile')) ? session('noty.config.transitionInMobile') : config('laravel-noty.transitionInMobile'))  }}',
				transitionOutMobile: '{{ (!empty(session('noty.config.transitionOutMobile')) ? session('noty.config.transitionOutMobile') : config('laravel-noty.transitionOutMobile'))  }}',
			});

            @foreach (session('noty.messages') as $item)
				iziToast.show({
					message: '{{ $item['text'] }}',
					color: '{{ $item['type'] }}',
					title: '{{ $item['title'] }}'
				});
            @endforeach
        })();
    </script>

	{{ session()->forget('noty.messages') }}
@endif


config/noty.php
<?php

return [
    /*
    * Title configuration color
    */
    'titleColor' => '',

    /*
    * Title configuration size
    */
    'titleSize' => '',

    /*
    * Title configuration line hight
    */
    'titleLineHeight' => '',

    /*
    * Message configuration color
    */
    'messageColor' => '',

    /*
    * Message configuration size
    */
    'messageSize' => '',

    /*
    * Message configuration line hight
    */
    'messageLineHeight' => '',

    /*
     * Positions:  bottomRight, bottomLeft, topRight, topLeft, topCenter, bottomCenter, center
     */
    'position' => 'topRight',

    /*
     * Themes:  light, dark
     */
    'theme' => 'light',

    /*
     * Layout:  1 (title inline with message), 2 (title on top of message)
     */
    'layout' => 1,

    /*
    * [integer|boolean] - Delay for closing event in milliseconds. Set false for sticky notifications.
    */
    'timeout' => 4000,

    /*
     * [boolean] - Displays a progress bar.
     */
    'progressBar' => true,

    /*
     * [string] - Progress bar color
     */
    'progressBarColor' => 'rgb(0, 255, 184)',

    /*
     * [string] - Progress bar linear / ease
     */
    'progressBarEasing' => 'linear',

    /*
     * [boolean] - Balloon
     */
    'balloon' => true,

    /*
     * [boolean] - enable to close?
     */
    'close' => true,

    /*
     * [boolean] - Close on click
     */
    'closeOnClick' => false,

    /*
     * [boolean] - Close on escape?
     */
    'closeOnEscape' => false,

    /*
     * Animate inside
     */
    'animateInside' => true, // bool

    /*
     * Enable drag for close
     */
    'drag' => true, // bool

    /*
     * Pause on hover
     */
    'pauseOnHover' => true, // bool

    /*
     * Reset on hover
     */
    'resetOnHover' => false, // bool

    /*
     * [string] - Transition In
     * Default toast open animation. It can be: bounceInLeft, bounceInRight, bounceInUp, bounceInDown, fadeIn, fadeInDown, fadeInUp, fadeInLeft, fadeInRight or flipInX.
     */
    'transitionIn' => 'fadeInUp',

    /*
     * [string] - Transition Out
     * Default toast close animation. It can be: fadeOut, fadeOutUp, fadeOutDown, fadeOutLeft, fadeOutRight, flipOutX
     */
    'transitionOut' => 'fadeOut',

    /*
     * [string] - Transition In Mobile
     */
    'transitionInMobile' => 'fadeInUp',

    /*
     * [string] - Transition Out Mobile
     */
    'transitionOutMobile' => 'fadeOutDown',
];
